File: app/src/main/java/com/khaled/move/navigation/foot/route/GeoUtils.kt
package com.khaled.move.navigation.foot.route

import com.khaled.move.navigation.foot.location.NavigationLocation
import com.khaled.move.navigation.foot.route.RoutePoint
import kotlin.math.PI
import kotlin.math.abs
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sin
import kotlin.math.sqrt

object GeoUtils {
    private const val EarthRadiusMeters = 6371000.0

    fun distanceMeters(a: RoutePoint, b: RoutePoint): Double =
        distanceMeters(a.latitude, a.longitude, b.latitude, b.longitude)

    fun distanceMeters(a: NavigationLocation, b: NavigationLocation): Double =
        distanceMeters(a.latitude, a.longitude, b.latitude, b.longitude)

    fun distanceMeters(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Double {
        val dLat = Math.toRadians(lat2 - lat1)
        val dLon = Math.toRadians(lon2 - lon1)
        val rLat1 = Math.toRadians(lat1)
        val rLat2 = Math.toRadians(lat2)
        val a = sin(dLat / 2) * sin(dLat / 2) +
                sin(dLon / 2) * sin(dLon / 2) * cos(rLat1) * cos(rLat2)
        val c = 2 * atan2(sqrt(a), sqrt(1 - a))
        return EarthRadiusMeters * c
    }

    fun bearingDegrees(from: RoutePoint, to: RoutePoint): Double {
        val lat1 = Math.toRadians(from.latitude)
        val lat2 = Math.toRadians(to.latitude)
        val dLon = Math.toRadians(to.longitude - from.longitude)
        val y = sin(dLon) * cos(lat2)
        val x = cos(lat1) * sin(lat2) - sin(lat1) * cos(lat2) * cos(dLon)
        return ((Math.toDegrees(atan2(y, x)) + 360.0) % 360.0)
    }

    fun angularDifferenceDegrees(first: Double, second: Double): Double {
        val diff = abs(first - second) % 360.0
        return if (diff > 180.0) 360.0 - diff else diff
    }

    fun normalizeBearingDegrees(value: Double): Double = ((value % 360.0) + 360.0) % 360.0

    fun shortestAngleDeltaDegrees(from: Double, to: Double): Double {
        val delta = normalizeBearingDegrees(to) - normalizeBearingDegrees(from)
        return when {
            delta > 180.0 -> delta - 360.0
            delta < -180.0 -> delta + 360.0
            else -> delta
        }
    }

    fun interpolateBearingDegrees(from: Double, to: Double, fraction: Double): Double =
        normalizeBearingDegrees(
            normalizeBearingDegrees(from) + shortestAngleDeltaDegrees(from, to) * fraction.coerceIn(
                0.0,
                1.0
            )
        )

    fun interpolate(a: RoutePoint, b: RoutePoint, fraction: Double): RoutePoint = RoutePoint(
        latitude = a.latitude + (b.latitude - a.latitude) * fraction,
        longitude = a.longitude + (b.longitude - a.longitude) * fraction,
    )

    fun destinationPoint(start: RoutePoint, bearingDegrees: Double, distanceMeters: Double): RoutePoint {
        val angularDistance = distanceMeters / EarthRadiusMeters
        val bearing = Math.toRadians(bearingDegrees)
        val lat1 = Math.toRadians(start.latitude)
        val lon1 = Math.toRadians(start.longitude)
        val lat2 = kotlin.math.asin(
            sin(lat1) * cos(angularDistance) + cos(lat1) * sin(angularDistance) * cos(bearing)
        )
        val lon2 = lon1 + atan2(
            sin(bearing) * sin(angularDistance) * cos(lat1),
            cos(angularDistance) - sin(lat1) * sin(lat2),
        )
        return RoutePoint(
            latitude = Math.toDegrees(lat2),
            longitude = Math.toDegrees(lon2),
        )
    }

    fun snapToSegment(point: RoutePoint, start: RoutePoint, end: RoutePoint): SnappedPoint {
        val scaleLat = 111320.0
        val scaleLon =
            cos(Math.toRadians((start.latitude + end.latitude + point.latitude) / 3.0)).coerceAtLeast(0.01) * 111320.0
        val px = point.longitude * scaleLon
        val py = point.latitude * scaleLat
        val ax = start.longitude * scaleLon
        val ay = start.latitude * scaleLat
        val bx = end.longitude * scaleLon
        val by = end.latitude * scaleLat
        val abx = bx - ax
        val aby = by - ay
        val apx = px - ax
        val apy = py - ay
        val abLengthSquared = abx * abx + aby * aby
        val t = if (abLengthSquared == 0.0) 0.0 else (apx * abx + apy * aby) / abLengthSquared
        val clamped = min(1.0, max(0.0, t))
        val snappedX = ax + abx * clamped
        val snappedY = ay + aby * clamped
        val snapped = RoutePoint(
            latitude = snappedY / scaleLat,
            longitude = snappedX / scaleLon,
        )
        return SnappedPoint(
            point = snapped,
            fraction = clamped,
            distanceMeters = distanceMeters(point, snapped),
        )
    }
}

data class SnappedPoint(
    val point: RoutePoint,
    val fraction: Double,
    val distanceMeters: Double,
)

File: app/src/main/cpp/utils/GeoUtils.cpp
#include "GeoUtils.h"
#include <cmath>
#include <algorithm>

namespace move {

double GeoUtils::toRadians(double degrees) {
    return degrees * M_PI / 180.0;
}

double GeoUtils::toDegrees(double radians) {
    return radians * 180.0 / M_PI;
}

double GeoUtils::distanceMeters(const RoutePoint& a, const RoutePoint& b) {
    return distanceMeters(a.latitude, a.longitude, b.latitude, b.longitude);
}

double GeoUtils::distanceMeters(double lat1, double lon1, double lat2, double lon2) {
    double dLat = toRadians(lat2 - lat1);
    double dLon = toRadians(lon2 - lon1);
    double rLat1 = toRadians(lat1);
    double rLat2 = toRadians(lat2);

    double a = std::sin(dLat / 2) * std::sin(dLat / 2) +
               std::sin(dLon / 2) * std::sin(dLon / 2) * std::cos(rLat1) * std::cos(rLat2);
    double c = 2 * std::atan2(std::sqrt(a), std::sqrt(1 - a));
    return EarthRadiusMeters * c;
}

double GeoUtils::bearingDegrees(const RoutePoint& from, const RoutePoint& to) {
    double lat1 = toRadians(from.latitude);
    double lat2 = toRadians(to.latitude);
    double dLon = toRadians(to.longitude - from.longitude);

    double y = std::sin(dLon) * std::cos(lat2);
    double x = std::cos(lat1) * std::sin(lat2) - std::sin(lat1) * std::cos(lat2) * std::cos(dLon);
    return normalizeBearingDegrees(toDegrees(std::atan2(y, x)));
}

double GeoUtils::normalizeBearingDegrees(double value) {
    double res = std::fmod(value, 360.0);
    if (res < 0) res += 360.0;
    return res;
}

double GeoUtils::shortestAngleDeltaDegrees(double from, double to) {
    double delta = normalizeBearingDegrees(to) - normalizeBearingDegrees(from);
    if (delta > 180.0) return delta - 360.0;
    if (delta < -180.0) return delta + 360.0;
    return delta;
}

RoutePoint GeoUtils::interpolate(const RoutePoint& a, const RoutePoint& b, double fraction) {
    return {
        a.latitude + (b.latitude - a.latitude) * fraction,
        a.longitude + (b.longitude - a.longitude) * fraction
    };
}

SnappedPoint GeoUtils::snapToSegment(const RoutePoint& point, const RoutePoint& start, const RoutePoint& end) {
    double scaleLat = 111320.0;
    double scaleLon = std::max(0.01, std::cos(toRadians((start.latitude + end.latitude + point.latitude) / 3.0))) * 111320.0;

    double px = point.longitude * scaleLon;
    double py = point.latitude * scaleLat;
    double ax = start.longitude * scaleLon;
    double ay = start.latitude * scaleLat;
    double bx = end.longitude * scaleLon;
    double by = end.latitude * scaleLat;

    double abx = bx - ax;
    double aby = by - ay;
    double apx = px - ax;
    double apy = py - ay;

    double abLengthSquared = abx * abx + aby * aby;
    double t = (abLengthSquared == 0.0) ? 0.0 : (apx * abx + apy * aby) / abLengthSquared;
    double clamped = std::max(0.0, std::min(1.0, t));

    double snappedX = ax + abx * clamped;
    double snappedY = ay + aby * clamped;

    RoutePoint snapped = { snappedY / scaleLat, snappedX / scaleLon };
    return { snapped, clamped, distanceMeters(point, snapped) };
}

} // namespace move
